Fix stats level for new counters

gaborznagy opened the pull request:
Stats level increased to level 1 for following new counters:
discarded, filter matched/not_matched, global allocated bytes

main function is also affected:
global allocated bytes stats initialization moved out from log_msg_global_init to log_msg_stats_global_init function.
It is needed as counter registration is configuration dependent when stats level is used (and in app_startup step the config is not yet parsed).
(See stats_check_level)
By default stats level 0 is working without a cfg, because the stat level check function allows this for stats_level 0.

kira-syslogng commented:
Build FAILURE, the tests were executed on test branch: master and test suite: functions

pzoleex commented:
There are two issues:
written counter is still displayed at stats_level(0)
embedded filters counter becomes wrong
I have this configuration:
@version: 7.0
options {
    stats_level(1);
};
source s_network_cf266a4f7cc0444bad2833c2d615aeb6 {
    network(ip("127.0.0.1")
 port(61010));
};
filter f_filter_8cd11f42c02d47998ebb4c28b330587c {
    facility(mail);
};
filter f_filter_15aa5caf50464cb8a0497126c369b52d {
    facility(cron);
};
filter f_filter_a5b0559526954d93ac8ba74673caeb96 {
    not filter(f_filter_15aa5caf50464cb8a0497126c369b52d) and not filter(f_filter_8cd11f42c02d47998ebb4c28b330587c);
};
destination d_file_0c474b00c4384d4da1ff849db4eabf2f {
    file("/var/testdb_working_dir/e67315db-82f0-43f6-8cdc-81321c9922e2.txt");
};
destination d_file_8c7184b0779e4b56b1b6d5d7b53f92f7 {
    file("/var/testdb_working_dir/1b12ab9c-d5bb-44c8-a264-1fd6208e01d5.txt");
};
log {
source(s_network_cf266a4f7cc0444bad2833c2d615aeb6);
filter(f_filter_15aa5caf50464cb8a0497126c369b52d);
destination(d_file_0c474b00c4384d4da1ff849db4eabf2f);

flags(flow-control);
};log {
source(s_network_cf266a4f7cc0444bad2833c2d615aeb6);
filter(f_filter_a5b0559526954d93ac8ba74673caeb96);
destination(d_file_8c7184b0779e4b56b1b6d5d7b53f92f7);

flags(flow-control);
};

I send this line to syslog-ng:
<72>Feb 27 10:09:36 init: cron user

the expected output of this:
/opt/syslog-ng/sbin/syslog-ng-ctl query get "filter*f_filter_a5b0559526954d93ac8ba74673caeb96***.not_matched"

should be 2. After this patch it is 1.

pzoleex commented:
@kira-syslogng retest this please test branch=pzolee-stats-use-new-counters-stats-level-1;

kira-syslogng commented:
Build FAILURE, the tests were executed on test branch: pzolee-stats-use-new-counters-stats-level-1 and test suite: functions

lbudai commented on syslog-ng/main.c:
```diff
@@ -270,6 +270,7 @@ main(int argc, char *argv[])
   app_startup();
   main_loop_init(main_loop, &main_loop_options);
   rc = main_loop_read_and_init_config(main_loop);
+  log_msg_stats_global_init();
```
What if we could create a `register_stats_level_dependent_global_counters` method?
(and check all the global `init` functions that are registers counters... and move to there the registration?)

gaborznagy commented:
Updated Pzolee's comment for expected value after discussing the test.
Added : 
> the expected output of this: 
> /opt/syslog-ng/sbin/syslog-ng-ctl query get "filter.f_filter_15aa5caf50464cb8a0497126c369b52d.matched"

kira-syslogng commented:
Build FAILURE, the tests were executed on test branch: master and test suite: functions

pzoleex commented:
@kira-syslogng retest this please test branch=pzolee-stats-use-new-counters-stats-level-1;

kira-syslogng commented:
Build FAILURE, the tests were executed on test branch: master and test suite: functions

kira-syslogng commented:
Build FAILURE, the tests were executed on test branch: pzolee-stats-use-new-counters-stats-level-1 and test suite: functions

pzoleex commented:
@kira-syslogng retest this please test branch=pzolee-stats-use-new-counters-stats-level-1;

kira-syslogng commented:
Build SUCCESS, the tests were executed on test branch: pzolee-stats-use-new-counters-stats-level-1 and test suite: functions
